runtime/protocol/http-client-engines/http-client-engine-ktor/common/src/aws/smithy/kotlin/runtime/http/engine/ktor/KtorEngine.kt

```diff
@@ -4,15 +4,157 @@
  */
 package aws.smithy.kotlin.runtime.http.engine.ktor
 
-import aws.smithy.kotlin.runtime.http.engine.HttpClientEngine
-import aws.smithy.kotlin.runtime.http.engine.HttpClientEngineBase
-import aws.smithy.kotlin.runtime.http.engine.HttpClientEngineConfig
+import aws.smithy.kotlin.runtime.http.Headers
+import aws.smithy.kotlin.runtime.http.HttpStatusCode
+import aws.smithy.kotlin.runtime.http.engine.*
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.response.HttpCall
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.logging.Logger
+import aws.smithy.kotlin.runtime.logging.trace
+import aws.smithy.kotlin.runtime.time.Instant
+import io.ktor.client.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.util.*
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.job
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlin.coroutines.CoroutineContext
+import io.ktor.client.engine.HttpClientEngine as KtorHttpClientEngine
 
 /**
- * Specifies the ktor http client of which platform specific [HttpClientEngine]'s actualize
- *
- * @param config Provides configuration for the DefaultHttpClientEngine
+ * Utility class that wraps the given Ktor engine as an [HttpClientEngine].
+ * This class can be used to wrap any Ktor compliant engine (though not all engines
+ * may support HTTP features required by any given SDK).
  */
-expect class KtorEngine(config: HttpClientEngineConfig = HttpClientEngineConfig.Default) : HttpClientEngineBase {
-    val config: HttpClientEngineConfig
+class KtorEngine(
+    private val engine: KtorHttpClientEngine
+) : HttpClientEngineBase("ktor") {
+
+    @Suppress("UNUSED_PARAMETER")
+    @Deprecated(
+        message = "KtorEngine was previously synonymous with the OkHttp engine. It has been modified to wrap any " +
+            "Ktor compliant engine. The default engine has been changed from CRT to Ktor/OkHttp. To fix either " +
+            "remove setting `httpClientEngine` explicitly or instantiate a Ktor compliant engine of your own and " +
+            "use KtorEngine to wrap it. This constructor will be removed in a future release before GA.",
+        level = DeprecationLevel.ERROR
+    )
+    constructor(config: HttpClientEngineConfig = HttpClientEngineConfig.Default) : this(DeprecationEngine)
```

Comment:
**Comment**: It seems cleaner to throw an error immediately upon invoking `KtorEngine()` rather than later:

```kotlin
companion object {
    @Suppress("UNUSED_PARAMETER")
    @Deprecated(message = "KtorEngine was previously...", level = DeprecationLevel.ERROR)
    operator fun invoke(config: HttpClientEngineConfig = HttpClientEngineConfig.Default): KtorEngine =
        error("Invoking KtorEngine without an engine is deprecated")
}
```

Reply:
It's irrelevant, code doesn't compile if you target this constructor